### DISCUSSION TITLE
[Build] Refactor the build system(重构构建系统)

### DISCUSSION THREAD

**Discussion by mritd** [cloudreve/cloudreve#1607] [Ideas]
Currently we use build.sh to build distributions, but it still has some limitations, such as build dependencies, cache handling, platform-dependent variable handling, etc.

Can we use Taskfile to standardize the build system? For example, to unify our distribution file names, provide easy-to-understand cross-compilation configuration, etc.

---

目前来说我们的构建体系依赖于 build.sh 提供, shell 脚本非常灵活这是不可否认的; 不过对于一些缓存处理、任务依赖、交叉编译等仍然具有一些局限性.

我们是否可以考虑采用 Taskfile 构建工具来标准化构建体系? 比如统一文件名、Dockerfile、GitHub Action 乃至本地开发环境, 我个人认为 Taskfile 的好处是可以将构建操作通过配置文件的方式呈现出来, 并且可能会提高可靠性和可维护性.

- [go-task](https://taskfile.dev/)
- [Taskfile](https://taskfile.dev/usage/)
- [hysteria project example taskfile](https://github.com/apernet/hysteria/blob/master/Taskfile.yaml)
- [PR WIP](https://github.com/cloudreve/Cloudreve/pull/1608)

**Comment by HFO4** [comment]
Looks good to me!

**Comment by mritd** [comment]
At the moment I can think of some basic needs:

- Frontend or backend can be built separately (development environment or Docker multi-stage build)
- Support task dependencies, automatically detect preconditions for building tasks
- Support cache processing, shorten useless time-consuming
- Uniformly publish object file name format, which is beneficial for third-party installation scripts to download and update

About the CC compiler:

Is it necessary to define a CC compiler? If we only target the development environment and CI environment, I don't think it is necessary.

- For the development environment, the developer is theoretically aware of the CC compiler that needs to be used.
- For CI environments such as GitHub Action, we can force cross-compilation to be done using a specific environment

---

目前我能想到一些基本需求:

- 可单独构建前端或后端(开发环境或 Docker 多阶段构建)
- 支持任务依赖, 自动检测构建任务的前置条件
- 支持缓存处理, 缩短无用耗时
- 统一发布物物文件名称格式, 有利于第三方安装脚本下载更新

关于 CC 编译器:

是否有必要定义 CC 编译器? 如果我们只面向 开发环境和 CI 环境, 我认为似乎没有必要.

- 对于开发环境, 开发人员理论上很清楚需要使用的 CC 编译器.
- 对于 CI 环境例如 GitHub Action, 我们可以强制使用特定环境完成交叉编译
